darshan-util/darshan-logutils.h

```diff
@@ -273,4 +290,87 @@ int darshan_log_get_record(darshan_fd fd, int mod_idx, void **buf);
     memcpy(__ptr, __dst_char, 4); \
 } while(0)
 
+/*****************************************************************
+ * The functions in this section make up the accumulator API, which is a
+ * mechanism for aggregating records to produce derived metrics and
+ * summaries.
+ */
+
+/* opaque accumulator reference */
+struct darshan_accumulator_st;
+typedef struct darshan_accumulator_st* darshan_accumulator;
+
+/* Instantiate a stateful accumulator for a particular module type.
+ */
+int darshan_accumulator_create(darshan_module_id id,
+                               int64_t job_nprocs,
+                               darshan_accumulator*   new_accumulator);
+
+/* Add a record to the accumulator.  The record is an untyped void* (size
+ * implied by record type) following the convention of other logutils
+ * functions.  Multiple records may be injected at once by setting
+ * record_count > 1 and packing records into a contiguous memory region.
+ */
+int darshan_accumulator_inject(darshan_accumulator accumulator,
+                               void*               record_array,
+                               int                 record_count);
+
+struct darshan_file_category_counters {
+    int64_t count;                   /* number of files in this category */
+    int64_t total_read_volume_bytes; /* total read traffic volume */
+    int64_t total_write_volume_bytes;/* total write traffic volume */
+    int64_t max_read_volume_bytes;   /* maximum read traffic volume to 1 file */
+    int64_t max_write_volume_bytes;  /* maximum write traffic volume to 1 file */
+    int64_t total_max_offset_bytes;  /* summation of max_offsets */
+    int64_t max_offset_bytes;        /* largest max_offset */
+    int64_t nprocs;                  /* how many procs accessed (-1 for "all") */
+};
+
+enum darshan_file_category {
+    DARSHAN_ALL_FILES = 0,
+    DARSHAN_RO_FILES,
+    DARSHAN_WO_FILES,
+    DARSHAN_RW_FILES,
+    DARSHAN_UNIQ_FILES,
+    DARSHAN_SHARED_FILES,
+    DARSHAN_PART_SHARED_FILES,
+    DARSHAN_FILE_CATEGORY_MAX
+};
+
+/* aggregate metrics that can be derived from an accumulator.  If a given
+ * record time doesn't support a particular field, then it will be set to -1
+ * (for int64_t values) or NAN (for double values).
+ */
+struct darshan_derived_metrics {
+    int64_t total_bytes;                /* total bytes moved (read and write) */
+    double  slowest_rank_io_total_time; /* combined meta and rw time spent in
+                                           unique files on the slowest rank */
+    double slowest_rank_rw_only_time;   /* rw time spent in unique files on the
+                                           slowest rank */
+    double slowest_rank_md_only_time;   /* meta time spent in unique files on
+                                           the slowest rank */
+    int    slowest_rank_rank; /* which rank was the slowest for unique files */
```

Comment:
Maybe it's just me, but we might want to take the opportunity to clear up derived metric names here. Below, we have `shared_io_total_time_by_slowest` which I think makes sense but it's hard to visually compare to the "slowest" metrics above.

Maybe `unique_io_total_time_by_slowest` (and similar for other unique time measures) and just `slowest_rank`? I think that makes it easier when reading the code and see that these 2 values with disjointed names are being summed up ultimately (perf being calculated using the sum of unique+shared file times).

Reply:
Yeah, totally agreed.  Updated in 2cdb2a48..b526abcd.

Reply:
Looks great, thanks!